core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -556,6 +562,46 @@ public void copyLogSegmentsToRemote(UnifiedLog log) throws InterruptedException
             }
         }
 
+        public void cleanupDeletedRemoteLogSegments() {
+            if (isCancelled())
+                return;
+
+            Uuid topicId = topicIdPartition.topicId();
+            if (deletedTopicIds.contains(topicId)) {
+                cleanupAllRemoteLogSegments();
+                cancelRLMtask();
+                deletedTopicIds.remove(topicId);
+            }
+        }
+
+        private void cleanupAllRemoteLogSegments() {
```

Comment:
Since this runs asynchronously after topic deletion completes, if every replica is restarted before all remote segments are deleted, we will never be able to remove the remaining remote segments for the deleted topics?

Reply:
I think the topic deletion process of Kafka is irreversible.  In what circumstances would each replica restart after the deletion of a topic?

Reply:
This is the case that I am thinking about. Topic is being deleted and the controller marks the topic as deleted. The remote segment deletion is in progress but not completed. There is power outage and the whole cluster is down. After the cluster is restarted, since topic deletion won't be triggered again, the remaining remote segments for the deleted topic won't be cleaned. The local storage seems to have a similar issue right now since it's also deleted asynchronously. I am just wondering how we plan to address this new issue.

Reply:
Perhaps we can add a `RemoteLogCleaner` thread that periodically checks if there are expired/deleted segments on the remote.

Reply:
We could, but it adds its own overhead.